src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs

```diff
@@ -644,6 +644,88 @@ protected override void ApplyAnalyzerReferenceRemoved(ProjectId projectId, Analy
+            var mappedSpansPerDocument = GetTextChangesAndMappedSpans(projectChanges, changedMappedDocuments);
+
+            // Create mapped text changes using the mapped spans and original text changes.
+            using var _ = ArrayBuilder<(string FileName, TextChange TextChange)>.GetInstance(out var mappedChangesBuilder);
+            foreach (var mappedSpans in mappedSpansPerDocument)
+            {
+                var newDoc = projectChanges.NewProject.GetDocument(mappedSpans.Key)!;
+                MapTextChanges(newDoc, mappedSpans.Value.MappedSpans, mappedSpans.Value.OriginalTextChanges, mappedChangesBuilder);
+            }
+
+            var mappedChanges = mappedChangesBuilder.ToImmutable();
+            if (mappedChanges.Any())
+            {
+                // Group the mapped text changes by file, then apply all mapped text changes for the file.
+                foreach (var changesForFile in mappedChanges.GroupBy(change => change.FileName))
+                {
+                    using var invisibleEditor = new InvisibleEditor(ServiceProvider.GlobalProvider, changesForFile.Key, GetHierarchy(projectChanges.ProjectId), needsSave: true, needsUndoDisabled: false);
+                    TextEditApplication.UpdateText(changesForFile.Select(change => change.TextChange).ToImmutableArray(), invisibleEditor.TextBuffer, EditOptions.None);
+                }
+            }
+
+            static Dictionary<DocumentId, (ImmutableArray<MappedSpanResult> MappedSpans, ImmutableArray<TextChange> OriginalTextChanges)> GetTextChangesAndMappedSpans(
+                ProjectChanges projectChanges, ImmutableArray<DocumentId> changedMappedDocuments)
+            {
+                return Task.Run(async () =>
+                {
+                    var mappedSpansPerDocument = new Dictionary<DocumentId, (ImmutableArray<MappedSpanResult> MappedSpans, ImmutableArray<TextChange> OriginalTextChanges)>();
+                    foreach (var changedDocumentId in changedMappedDocuments)
+                    {
+                        var oldDocument = projectChanges.OldProject.GetDocument(changedDocumentId)!;
+                        var newDocument = projectChanges.NewProject.GetDocument(changedDocumentId)!;
+
+                        var mappingService = newDocument.Services.GetService<ISpanMappingService>();
+                        var textChanges = (await newDocument.GetTextChangesAsync(oldDocument, CancellationToken.None).ConfigureAwait(false)).ToImmutableArray();
+                        var mappedSpanResults = await mappingService.MapSpansAsync(oldDocument, textChanges.Select(tc => tc.Span), CancellationToken.None).ConfigureAwait(false);
+
+                        mappedSpansPerDocument[changedDocumentId] = (mappedSpanResults, textChanges);
+                    }
+
+                    return mappedSpansPerDocument;
+                }).WaitAndGetResult(CancellationToken.None);
```

Comment:
you don't need to explicitly kick of a Task onto anotehr thread here.  you can just have this be a helper function then call that helper function and Wait on that result.

Reply:
@jasonmalinowski i really hate that we effectively have this sync api asking clients to tell us what to do, but we need to then call async services, that we synchronously block on to *actually* determine what to do.  should this mapping and whatnot happen prior to ApplyProjectChanges being called in the first place?

Reply:
Note that I could move this up to the inlinerenamesession - http://sourceroslyn.io/#Microsoft.CodeAnalysis.EditorFeatures.Wpf/InlineRename/InlineRenameSession.cs,743630c69dd0288f but the commit operation on the rename is entirely synchronous anyway.  

So to really move it to an async point I'd need to move it up to maybe here - http://sourceroslyn.io/#Microsoft.CodeAnalysis.EditorFeatures.Wpf/InlineRename/AbstractEditorInlineRenameService.SymbolRenameInfo.cs,189 This is where the locations are generated (and not just for commit, but we'd be calling the mapping service on every char typed).  I could then store those in the inlinerenameset and manually apply those after workspace.TryApplyChanges. This would not just work for rename tracking as tryapplychanges wouldn't have that.  Open to ideas here

Reply:
So I think this is the right place to put this, since putting it into inline rename code specifically means it can't be used for any other refactoring like say change signature. I presume we've already tested change signature with this PR and verified it's working because I think it will right now!

Should this all be async? If we had a time machine, then maybe we'd make TryApplyChanges be asynchronous, although we had various reasons at the time to not do so, and I'm not sure anything has changed anything there. And long term with LSP, this code goes away, so I'm not sure if I'd spend any time in trying to thread async through. And honestly, this here is still probably cheaper than opening of the invisible editor and all of the other _actually_ expensive stuff. :smile:

Reply:
@CyrusNajmabadi's about the odd Task.Run is still correct -- the WaitAndGetResults() can just be inlined, and if you're not passing in a cancellation token you can just .Result and be done with it.

Reply:
@jasonmalinowski @CyrusNajmabadi so I tried out doing `var mappedChanges = GetMappedTextChanges(projectChanges, changedMappedDocuments).Result;` instead of Task.Run - but this always seems to deadlock when calling the span mapping service, whereas `var mappedChanges = Task.Run(async () => await GetMappedTextChanges(projectChanges, changedMappedDocuments).ConfigureAwait(true)).WaitAndGetResult(CancellationToken.None);` does not.

I'm not 100% of the semantics here, as far as I understand it, the task.run will run the work on the threadpool thread  but .Result runs synchronously?  Could there be some issue here where razor is not expecting to be on the UI thread?

Reply:
Something else is up, I'd expect the Task.Run() version to deadlock _more_. WaitAndGetResult() is just .Result under the covers.

Reply:
Figured it out.  See https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1216657

I don't think we'll get a fix for the above in time for p4, so I'm going to leave as-is.  I can followup with a change once the above is resolved.